src/ray/raylet/worker_pool.cc

```diff
@@ -711,17 +714,28 @@ void WorkerPool::TryKillingIdleWorkers() {
       auto rpc_client = worker->rpc_client();
       RAY_CHECK(rpc_client);
       rpc::ExitRequest request;
-      rpc_client->Exit(request, [](const ray::Status &status, const rpc::ExitReply &r) {
-        if (!status.ok()) {
-          RAY_LOG(ERROR) << "Failed to send exit request: " << status.ToString();
-        }
-      });
-      // Remove the worker from the idle pool so it can't be popped anymore.
-      RemoveWorker(worker_state.idle, worker);
-      if (!worker->IsDead()) {
-        worker->MarkDead();
-        running_size--;
-      }
+      rpc_client->Exit(
+          request, [this, worker](const ray::Status &status, const rpc::ExitReply &r) {
+            if (!status.ok()) {
+              RAY_LOG(ERROR) << "Failed to send exit request: " << status.ToString();
+              return;
+            }
+
+            if (r.success()) {
+              auto &worker_state = GetStateForLanguage(worker->GetLanguage());
+              // If we could kill the worker properly, we remove them from the idle pool.
+              if (RemoveWorker(worker_state.idle, worker)) {
+                // If the worker is not idle at this moment, we don't mark them dead.
+                // It will be marked as dead after the current task is finished.
+                if (!worker->IsDead()) {
+                  worker->MarkDead();
+                }
+              }
+            }
+          });
+      // Always subtract the running size here so that it avoids killing too many io
+      // workers unncessarily.
+      running_size--;
```

Comment:
What if the "Exit" rpc request took more than 200ms to reach the worker? I guess it's possible that a worker is "exited" twice and the `running_size` counter is wrong.

Reply:
Subtracting the running size here is necessary because without that, we will always kill every idle worker (because we keep killing workers until the running size is smaller than the soft limit).

The scenario you mentioned is actually not only the issue when Exit is received within 200ms. We can always have the wrong running_size here because we don't know if the idle worker will actually be killed until we received a reply.

So let me explain why this is harmless instead (having wrong running_size). The worst case scenario here is we would have "lower"  running_size than it is supposed to be. The worst possible issue here is we will kill less idle workers than we needed to for 200 ms (which is fine because eventually idle workers will be killed properly).

Reply:
I agree with @kfstorm this is kind of weird. How about we just keep the size as is? The raylet will keep trying to kill workers every 200ms, but that's fine, it doesn't add much overhead.

Reply:
How about we keep two counters:
running_size, and exit_pending_size?

That way this for loop will still work, but we still have accurate counts.

Reply:
`running_size` is not a private attribute. So, technically what running_size means now is `exit_pending_size`. What about I just rename the variable?

Reply:
I see, that seems ok then. We might want to change the iteration order here to be random instead of always the same order though.

Reply:
Seems like there are some concerns around the correctness (including the random access feedback), so I just made a new map to keep tracking of pending exit workers. 

- We subtract pending exit workers from the running size in the beginning.
- If we see the worker that is pending exit, we continue in the loop. 

I think this will resolve the problem you wanted to solve by random access too. (and this will help us keeping FIFO order)

Reply:
Oh, my bad. I thought `running_size` is a private field, but it's a local variable. With is new pending exiting workers map, now the iteration order is not limited to FIFO, right? Cause without the map, we have to make sure a consistent iteration order.

Reply:
We still kill idle workers in FIFO order because now we subtract the pending exit map size from the running size in the beginning of the function. So, we stop killing idle workers if we have soft_limit + N idle workers and N of workers are pending exit. 

But I just realized this could have an issue if the first N workers own objects. (We won't kill other workers in the back of the queue in this case). Let me push the fix (I guess we should just go with the random order in this case...)

Reply:
Great catch! Let's make sure it's covered by UT. Be careful that don't kill too many workers, though.

Reply:
I think if we pop the front of the queue and push to the back whenever the idle worker is failed to be killed, we can handle this well without killing too many workers.